docs/gguf.md

```diff
@@ -0,0 +1,513 @@
+}
+
+// A string in GGUF.
+struct gguf_string_t {
+    // The length of the string, in bytes.
+    uint32_t len;
+    // The string as a UTF-8 non-null-terminated string.
+    char string[len];
+}
+
+union gguf_metadata_value_t {
+    uint8_t uint8;
+    int8_t int8;
+    uint16_t uint16;
+    int16_t int16;
+    uint32_t uint32;
+    int32_t int32;
+    float float32;
+    bool bool_;
+    gguf_string_t string;
+    struct {
+        // Number of elements, not bytes
+        uint32_t len;
+        // Any value type is valid, including arrays.
+        gguf_metadata_value_type type;
+        // The array of values.
+        gguf_metadata_value_t array[len];
+    } array;
+};
+
+struct gguf_metadata_kv_t {
+    // A standard GGUF string, with the following caveats:
+    // - It must be a valid ASCII string.
+    // - It must be a hierarchical key, where each segment is `lower_snake_case` and separated by a `.`.
+    // - It must be at most 2^16-1 bytes long.
+    // Any keys that do not follow these rules are invalid.
+    gguf_string_t key;
+
+    // The length of the value, in bytes
+    uint32_t value_len;
+    // The type of the value.
+    // Must be one of the `gguf_metadata_value_type` values.
+    gguf_metadata_value_type value_type;
+    // The value.
+    gguf_metadata_value_t value;
+};
+
+struct gguf_header_t {
+    // Magic number to announce that this is a GGUF file.
+    // Must be `'GGUF'`/`0x47475546`.
+    uint32_t magic;
+    // The version of the format implemented.
+    // Must be `1` for version described in this spec.
+    //
+    // This version should only be increased for structural changes to the format.
+    // Changes that do not affect the structure of the file should instead update the metadata
+    // to signify the change.
+    uint32_t version;
+    // The number of tensors in the file.
+    // This is explicit, instead of being included in the metadata, to ensure it is always present
+    // for loading the tensors.
+    uint32_t tensor_count;
+    // The number of metadata key-value pairs.
+    uint32_t metadata_kv_count;
+    // The metadata key-value pairs.
+    gguf_metadata_kv_t metadata_kv[metadata_kv_count];
+};
+
+struct gguf_tensor_info_t {
+    // The name of the tensor.
+    gguf_string_t name;
+    // The number of dimensions in the tensor.
+    // Currently at most 4, but this may change in the future.
+    uint32_t n_dimensions;
+    // The dimensions of the tensor.
+    uint32_t dimensions[n_dimensions];
+    // The number of elements in the tensor.
+    uint32_t n_elements;
+    // The type of the tensor.
+    ggml_type type;
+    // The offset of the tensor's data in this file in bytes.
+    // Must be a multiple of `ALIGNMENT`.
+    uint64_t offset;
+};
+
+struct gguf_file_t {
+    // The header of the file.
+    gguf_header_t header;
+
+    // Padding to the nearest multiple of `ALIGNMENT`.
+    uint8_t _padding[ALIGNMENT - (sizeof(header) % ALIGNMENT)];
```

Comment:
Shouldn't this padding be after `tensor_infos` and before `tensor_data`?

Reply:
Yes. My bad.